### PR TITLE
Companionable Circumstances

### PR DESCRIPTION
**Pull request by salt-die**

### CODE REVIEW

**Comment by lemonsaurus** [comment]
Lint failing:

```
./circumstantial-companions/main.py:138:26: F821 undefined name '_'
./circumstantial-companions/main.py:147:23: F821 undefined name '_'
./circumstantial-companions/main.py:153:23: F821 undefined name '_'
./circumstantial-companions/main.py:193:27: F821 undefined name '_'
./circumstantial-companions/main.py:214:22: F821 undefined name '_'
./circumstantial-companions/main.py:219:29: F821 undefined name '_'
./circumstantial-companions/main.py:222:29: F821 undefined name '_'
./circumstantial-companions/main.py:230:53: F821 undefined name '_'
./circumstantial-companions/main.py:237:30: F821 undefined name '_'
./circumstantial-companions/main.py:252:37: F821 undefined name '_'
./circumstantial-companions/main.py:253:38: F821 undefined name '_'
./circumstantial-companions/main.py:255:20: F821 undefined name '_'
./circumstantial-companions/main.py:273:46: F821 undefined name '_'
./circumstantial-companions/main.py:279:32: F821 undefined name '_'
./circumstantial-companions/main.py:323:22: F821 undefined name '_'
./circumstantial-companions/main.py:350:30: F821 undefined name '_'
./circumstantial-companions/main.py:357:30: F821 undefined name '_'
./circumstantial-companions/main.py:363:32: F821 undefined name '_'
./circumstantial-companions/main.py:400:9: E731 do not assign a lambda expression, use a def
./circumstantial-companions/main.py:401:9: E731 do not assign a lambda expression, use a def
./circumstantial-companions/main.py:403:9: E731 do not assign a lambda expression, use a def
./circumstantial-companions/main.py:404:9: E731 do not assign a lambda expression, use a def
./circumstantial-companions/main.py:437:28: F821 undefined name '_'
./circumstantial-companions/main.py:445:31: F821 undefined name '_'
./circumstantial-companions/main.py:452:29: F821 undefined name '_'
./circumstantial-companions/main.py:459:30: F821 undefined name '_'
./circumstantial-companions/main.py:466:28: F821 undefined name '_'
./circumstantial-companions/main.py:473:26: F821 undefined name '_'
./circumstantial-companions/main.py:518:32: F821 undefined name '_'
./circumstantial-companions/main.py:524:36: F821 undefined name '_'
./circumstantial-companions/main.py:535:1: E302 expected 2 blank lines, found 1
./circumstantial-companions/stone.py:7:25: E261 at least two spaces before inline comment
```

**Comment by SebastiaanZ** [comment]
## Code Review

**Commit Quality**

The overall quality of commits in this project is poor. A lot of the commit messages are non-descriptive (several say "initial" + things like "a line" and "lone parenthesis looks gross") or generic (e.g., "hooked up sliders", "WIP: Appearance changes"). None of the commit messages contain more than a single summary line. This makes it difficult to get an idea of what each commit does when scrolling through the commit history without studying the code changes themselves.

There are also a lot of commits (242) and they do not always correspond well to atomic changes: Sometimes related changes are spread out over several different commits, while other commits combine seemingly unrelated changes. As an example, the consecutive commits 0ea4407 ("app doc string") and e94d21b ("doc strings") both add in a single docstring and could probably have been combined; at the same time, the first one also changes a totally unrelated offset constant in a different file, which should probably have been part of another commit. In addition, there are a lot of commits that could have been prevented with the use of a proper pre-commit hook (e.g., commit such as "pep fixes", "indentation updated", and "removed newline").

To provide some background on what makes a good commit, here are some resources:

- https://fatbusinessman.com/2019/my-favourite-git-commit
- https://vimeo.com/172882423
- http://ablogaboutcode.com/2011/03/23/proper-git-commit-messages-and-an-elegant-git-history
- https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
- https://thoughtbot.com/blog/5-useful-tips-for-a-better-commit-message
- https://chris.bPeams.io/posts/git-commit/

**The Readme**

The README of the project is decent. It has a nice preview GIF showcasing the application in action and it's combined with a catchy general description. The installation and usage sections are basic; they just list the steps you need to take in order to install and run the app. There's no documentation on how to use the app itself and there's no mention of the extensive internationalization that was added to the application (other than that it's visible in the preview GIF). The attributions section is excellent and could serve as an example of how to attribute properly.

**Code Quality**

I like the use of descriptive variable names; it really helps with making the code self-documenting. I do think the code could benefit from more comments and docstrings, as there are whole sections without them. (As a side note, the reverse is also true: The `OptionsPanel.build` method contains a lot of redundant comments like `# Title` followed by `title = ...`.) The `mixins.py` file does have some excellent docstrings, so I suspect that the functional implementation took priority over documentation as the project progressed.

The code itself is generally readable, although it could use a bit of vertical spacing in particularly dense blocks to separate it out into smaller, logical groupings. It also contains plenty of idiomatic patterns that I like, such as using `pathlib` to join paths and context managers to manage opened file contexts. At the same time, it also contains various anti-patterns such as using the `or` operator solely for its side-effect of calling multiple callables that return `None` (e.g, `lambda: bg_func() or trans_func() or project_func()` in [this section](https://github.com/salt-die/code-jam-6/blob/da9975c3b9b33afc857d2dce8ca75a66035e919b/circumstantial-companions/main.py#L400-L411)). These type of things make the code difficult to follow for readers by hiding the intention of the programmer. The internationalization, on the other hand, is a really cool addition and is done in the proper, idiomatic way using `gettext`/`I18N`.

Overall, the code shows competence and knowledge of Python, but the style, documentation, and readability aspects could use a bit of attention. I do like the general object-oriented organization; it really does make the application more easy to reason about. The solutions to get around some of the limitations of the normal way of rendering in Kivy are also very creative.

Taken together, I'd say the code quality is average.

**Pros**
- Good variable names
- Use of idiomatic patterns
- The use of a logical, object-oriented structure
- Creative use of Kivy
- Excellent attributions for the assets

**Cons**
- Large parts of the code lack documentation (docstrings/comments)
- Use of anti-patterns (code golf-style) that obfuscate the intention of the programmer
- Poor commit quality

**Comment by aeros** [comment]
> At the same time, it also contains various anti-patterns such as using the or operator solely for its side-effect of calling multiple callables that return None (e.g, lambda: bg_func() or trans_func() or project_func() in this section).

I'd like to specifically draw emphasis to this point: I very highly recommend against using the `or` operating solely for the purpose of chaining together multiple function calls. It very much goes against the guiding philosophy of Python in having clear and obvious intention of what the code is meant to do. I'd also recommend against this as well: `self.chisel.on_touch_down = self.chisel.on_touch_move = lambda *args: None` (https://github.com/salt-die/code-jam-6/blob/da9975c3b9b33afc857d2dce8ca75a66035e919b/circumstantial-companions/main.py#L572).

But overall, the good organization and general layout of the code made it much easier to follow. Some additional docstrings that briefly explained the purpose of each class and function would have been helpful though, as well as some improved variable and function naming. A common mistake that I've seen in a few other submissions has been attempting to cram far too many unrelated tasks into a single function (poor coupling), which this project did a decent job of avoiding.

(Not a full review since I wasn't assigned to this one, just wanted add my thoughts on the above parts)

**Comment by salt-die** [comment]
I found no other method to disable chiseling while menu was open, but
monkey-patching.  Even the normal Kivy way of returning True from an
on_touch method to indicate a touch was handled didn't work.  So I have no
good idea how to avoid the patching to get the behavior we wanted.  I have
some bad ideas on how to do it though.  But above was simplest method.

I tend to agree with `or` thing with the caveat that we were implementing
quite a lot of polish under a heavy time constraint.

**Comment by aeros** [comment]
> I tend to agree with `or` thing with the caveat that we were implementing
quite a lot of polish under a heavy time constraint.

I can certainly understand having to sacrifice a bit of elegance for the sake of time constraints, but `all_func = lambda: bg_func() or trans_func() or project_func()` could have been just as easily written as:

```py
def all_func():
    bg_func()
    trans_func()
    project_func()
```

(as an inner function)

**Comment by MusicOnline** [comment]
> Terrible commit quality.

I'm perfectly aware of that, I guess we didn't really bother with good version control since it was just a quick and fun code jam product instead of a proper, serious project that would be maintained post-code-jam. (to the point where I pollute the git history with merge branch commits! yay!)

> I do think the code could benefit from more comments and docstrings, as there are whole sections without them.

I did not add them because I was the only one working on the front-end (appearance) side of things and salt was the only one working on the back-end (`stone.py`). As mentioned above and in your review, we took priority on implementation than documentation that no one would read (during the code jam or for serious purposes post-code-jam). That one single well documented mixin was just for fun.

> As a side note, the reverse is also true: The `OptionsPanel.build` method contains a lot of redundant comments like `# Title` followed by `title = ...`.

I like separating them into logical categories or so I think.

> At the same time, it also contains various anti-patterns such as using the or operator solely for its side-effect of calling multiple callables that return `None` (e.g, `lambda: bg_func() or trans_func() or project_func()` in this section).

Understood. I wasn't feeling like adding an entire function definition I guess :<
(seemed like a waste of vertical space at the time of writing)

> The most striking decision is the total avoidance of kv files.

When I first touched the project, I wanted to switch everything to kv. However, I could not get Babel to extract translatable strings from those kv files. That's why I completely abandoned kv files to keep everything in one place.

> The main.py is too big and should be split to smaller components.

Understood! My fault for rushing things, to the point where I did not have time to review that part of organization in the end. I would just place them in a `widgets/.__init__.py` and `widgets/popups.py` I guess.

> Good canvas management to animate many pebbles without performance problems. This isn't hard in itself, but new users sometimes struggle to find the best way to manage many instructions.

Thanks salt!

> The cursor drawing is a neat, self contained piece of code that works very nicely and is an excellent demonstration of how to achieve this type of effect.

Thanks david!

> Minor, but using an explicit relative path would be safer for people running the app in slightly different ways (i.e. with different working directories).
(And the same elsewhere in the app)

Which one are you pointing?

> Using `kivy.metrics.dp(40)` would be a little better for correct scaling on high dpi displays. As it is your cursor won't be a consistent size.

My fault for not changing that mid-development! Then again, I had to look up density-independent pixel and try to wrap my head around it because I do not have the privilege of high-density displays.

> I guess for perfect style these details should be filled in :<

It's not like anyone is going to maintain those™️.

> `I think self.canvas.ask_update()` should work as the more explicit way to trigger this, but the comment does make it clear what's going on.

I did not know that existed and that was a lucky bug fix :<

> Looks like this is dev code left in? Even if this has some value, it should be commented why!

For testing the Chisel by itself without the Options menu! (that is before I started working on it)

> Minor bug: the on_text_validate binding fires when pressing enter, which you can do before selecting a file type. This crashes the app.
(Found while trying the app, not with superhuman code reading skills)

My fault! I thought I tested it very intensely.

> This is an interesting pattern I haven't seen before - what is the motivation for wanting the initial Window.size to be a special minimum value? I can believe there's a good reason but it isn't familiar to me.

It is so that I don't need to consider the fact that the window can be resized to such a small size that everything might look horrendous.

If you've reached the end of this comment, thanks for reviewing our code!!!

**Comment by salt-die** [comment]
self.canvas.ask_update() doesn't seem to update the texture correctly for
what it's worth

File: circumstantial-companions/main.py [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+        locales = {code: info["name"] for code, info in LOCALES.items()}
+        popup = SelectionPopup(_("Select language"), locales)
+        popup.bind(choice=lambda instance, choice: self.build(choice))
+        popup.open()
+
+    def reset_chisel(self, *args):
+        Window.remove_widget(CURSOR)
+        popup = open_loading_popup(_("Resetting the canvas."))
+        Clock.schedule_once(lambda dt: (self.chisel.reset()
+                                        or popup.dismiss()
```
[the remark this answers is not in the document]
Comment (tshirtman) [reply circumstantial-companions/main.py]: and replacing the `or` by a simple `, ` would work just as well.

**Pull request merged**